unison-cli/src/Unison/Codebase/Editor/HandleInput.hs

```diff
@@ -2621,41 +2609,49 @@ showTodoOutput getPpe patch names0 = do
         .= ( Text.unpack . Reference.toText . view _2
                <$> fst (TO.todoFrontierDependents todo)
            )
-      ppe <- getPpe
+      -- only needs the local references to check for obsolete defs
+      ppe <- do
+        names <- makePrintNamesFromLabeled' (Patch.labeledDependencies patch)
+        prettyPrintEnvDecl names
       Cli.respondNumbered $ TodoOutput ppe todo
 
 checkTodo :: Patch -> Names -> Cli r (TO.TodoOutput Symbol Ann)
 checkTodo patch names0 = do
   Cli.Env {codebase} <- ask
-  let shouldUpdate = Names.contains names0
-  f <- Propagate.computeFrontier (liftIO . Codebase.dependents codebase) patch shouldUpdate
-  let dirty = R.dom f
-      frontier = R.ran f
-  (frontierTerms, frontierTypes) <- loadDisplayInfo frontier
+  let -- Get the dependents of a reference which:
+      --   1. Don't appear on the LHS of this patch
+      --   2. Have a name in this namespace
+      getDependents :: Reference -> IO (Set Reference)
+      getDependents ref = do
+        dependents <- Codebase.dependents codebase Queries.ExcludeSelf ref
+        pure (dependents & removeEditedThings & removeNamelessThings)
+  -- (r,r2) ∈ dependsOn if r depends on r2, excluding self-references (i.e. (r,r))
+  dependsOn <- liftIO (Monoid.foldMapM (\ref -> R.fromManyDom <$> getDependents ref <*> pure ref) edited)
+  let dirty = R.dom dependsOn
+  transitiveDirty <- liftIO (transitiveClosure getDependents dirty)
```

Comment:
Hrmm, what happens if I've lost a name in the middle of a transitive dep;

E.g. I edit 'foo', and it needs to propagate, but I've also got the dependency chain:

`x -> #abc -> foo`; It seems like because we filter out nameless things during discovery of the transitive closure, `x` would not be detected as a named dependent even though it is.

If the answer ends up being that it adds a bunch of complexity for a (hopefully) rare case, then that's probably fine, but nameless things seem to come up relatively regularly these days so I figure we should think about it it 😄

Reply:
Yeah, you won't see a `todo` entry in this case. This PR doesn't change that behavior, but I did notice that, too, and thought maybe it'd be better if we don't stop the search when we encounter a nameless thing. But, namespaces are supposed to be self-contained so I'm not sure.

Reply:
Namespaces are supposed to be self-contained but it's hard to enforce, so unless we have some namespace statistic that asserts that it _is_ self-contained (cc @ChrisPenner), we should probably keep looking. (In the next PR).